frontends/p4/duplicateActionControlPlaneNameCheck.h

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2024 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef FRONTENDS_P4_DUPLICATEACTIONCONTROLPLANENAMECHECK_H_
+#define FRONTENDS_P4_DUPLICATEACTIONCONTROLPLANENAMECHECK_H_
+
+#include <vector>
+
+#include "ir/ir.h"
+#include "ir/visitor.h"
+
+namespace P4 {
+
+/**
+ * This pass does not change anything in the IR.  It only gives an
+ * error if it finds two actions with the same hierarchical name.  I
+ * am not certain, but it might be that at this point in the front
+ * end, the only way such a duplicate can happen is due to @name
+ * annotations.  @name annotations are definitely taken into account
+ * in this duplicate check.
+ *
+ * See also the pass HierarchicalNames, on which the implementation of
+ * this pass was based.
+ *
+ * This pass should be run before pass LocalizeAllActions, because
+ * LocalizeAllActions can create actions with duplicate names, by
+ * design, that were not created by the P4 developer.  We should not
+ * issue an error if that pass creates duplicate hierarchical names.
+ */
+class DuplicateActionControlPlaneNameCheck : public Transform {
+    std::vector<cstring> stack;
+    /// Used for detection of conflicting control plane names among actions.
+    absl::flat_hash_map<cstring, const IR::Node *> actions;
+
+ public:
+    cstring getName(const IR::IDeclaration *decl);
+
+    DuplicateActionControlPlaneNameCheck() {
+        setName("DuplicateActionControlPlaneNameCheck");
+        visitDagOnce = false;
+    }
+    const IR::Node *preorder(IR::P4Parser *parser) override {
+        // There cannot be any action definitions inside a parser, so
+        // do not traverse through its nodes.
+        prune();
+        return parser;
+    }
+
+    const IR::Node *preorder(IR::P4Control *control) override {
+        stack.push_back(getName(control));
+        return control;
+    }
+    const IR::Node *postorder(IR::P4Control *control) override {
+        stack.pop_back();
+        return control;
+    }
```

Comment:
Instead of using `stack.empty()` to check whether we are processing a top-level action or not, it should be possible to replace `stack.empty()` with `!findContext<IR::P4Control>()` and remove all other uses of `stack`, including here.

If efficiency of this approach is of concern, then maybe consider using a simple counter (initialized to 0) which is incremented inside of `const IR::Node *preorder(IR::P4Control *control)`, and decremented inside of `const IR::Node *postorder(IR::P4Control *control)`. Then you could check whether `counter == 0` instead of `stack.empty()` to determine whether we are processing a top-level action.

Reply:
`stack`'s primary use is in the body of method `DuplicateActionControlPlaneNameCheck::postorder(IR::P4Action *action)`, to construct the hierarchical name of an action that has a relative name (i.e. one that does not begin with ".").  I don't see any way to eliminate `stack`.

The use of `stack.empty()` use to determine top-level or not is something that Oleg actually suggested using instead of my original `!findContext<IR::P4Control>()` :-). It wasn't a functionality change -- he guessed it might be slightly more efficient to do the `stack.empty()` call instead.  I have not tried to measure it.  I have a hard time believing it makes a measurable difference either way.

Reply:
Oh, sorry, I didn't see it used on this line at first:
```
            } else {
                name = absl::StrCat(".", absl::StrJoin(stack, "."), ".", name);
```
so ignore my suggestion :)